src/VisualStudio/Core/Def/Implementation/MoveToNamespace/VisualStudioMoveToNamespaceOptionsService.cs

```diff
@@ -2,45 +2,83 @@
 using System.Composition;
 using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.LanguageServices;
 using Microsoft.CodeAnalysis.MoveToNamespace;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.MoveToNamespace
 {
     [Export(typeof(IMoveToNamespaceOptionsService)), Shared]
     internal class VisualStudioMoveToNamespaceOptionsService : IMoveToNamespaceOptionsService
     {
+        private readonly string?[] _history;
+        private readonly Func<MoveToNamespaceDialogViewModel, bool?> _showDialog;
+
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public VisualStudioMoveToNamespaceOptionsService()
         {
+            _history = new string[3];
+            _showDialog = viewModel => new MoveToNamespaceDialog(viewModel).ShowModal();
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("RoslynDiagnosticsReliability", "RS0034:Exported parts should be marked with 'ImportingConstructorAttribute'", Justification = "Test constructor")]
+        internal VisualStudioMoveToNamespaceOptionsService(string?[] history, Func<MoveToNamespaceDialogViewModel, bool?> showDialog)
+        {
+            _history = history;
+            _showDialog = showDialog;
         }
 
         public MoveToNamespaceOptionsResult GetChangeNamespaceOptions(
             string defaultNamespace,
             ImmutableArray<string> availableNamespaces,
-            ISyntaxFactsService syntaxFactsService)
+            ISyntaxFacts syntaxFactsService)
         {
             var viewModel = new MoveToNamespaceDialogViewModel(
                 defaultNamespace,
                 availableNamespaces,
-                syntaxFactsService);
+                syntaxFactsService,
+                _history.WhereNotNull().ToImmutableArray());
 
-            var dialog = new MoveToNamespaceDialog(viewModel);
-            var result = dialog.ShowModal();
+            var result = _showDialog(viewModel);
 
             if (result == true)
             {
+                OnSelected(viewModel.NamespaceName);
                 return new MoveToNamespaceOptionsResult(viewModel.NamespaceName);
             }
             else
             {
                 return MoveToNamespaceOptionsResult.Cancelled;
             }
         }
+
+        private void OnSelected(string namespaceName)
+        {
+            var currentIndex = _history.IndexOf((n) => n == namespaceName);
+            if (currentIndex >= 0)
+            {
+                _history[currentIndex] = null;
+            }
+
+            for (var i = _history.Length - 1; i > 0; i--)
+            {
+                if (_history[i - 1] == null)
+                {
+                    continue;
+                }
+
+                _history[i] = _history[i - 1];
+            }
+
+            _history[0] = namespaceName;
+        }
```

Comment:
using a linked-list (or pretty much any other data structure) woudl make this much simpler :)

Reply:
this is a UI component witha  capped capacity.  There's no need to micro-optimize with an array.

Reply:
sure, changed to a LinkedList